### ISSUE TITLE
Tree shaking

### ISSUE THREAD

**Issue by birkskyum** [maplibre/maplibre-gl-js#977]
As I understand, tree shaking is not properly supported currently. We can make this a tracking issue and get an overview for what needs to happen in order to get it working.

**Comment by chanmathew** [comment]
hi @HarelM - just wondering if there has been any updates on this or if this is still currently blocked?

**Comment by birkskyum** [comment]
@chanmathew , I believe the latest update in this regards is that there not too long ago was a large refactor towards named imports, allowing unused pieces to be left out. 

We'd be able to improve our support for tree shaking by distributing an ESM bundle, but that requires all the dependencies to be updated from CJS to ESM too. There is a blocker here, related to updating the dependencies, because this one has some issues, and resolution is on hold due to a parental leave:
- https://github.com/maplibre/maplibre-gl-js/pull/4393

Another piece of the puzzle is tooling. Our test harness uses Jest, which only has [experimental ESM support](https://jestjs.io/docs/ecmascript-modules). Last time I tried running it with ESM, there were some issues for our bundle tests. I'm hoping that the upcoming [Jest 30](https://github.com/jestjs/jest/milestone/15) will improve on the situation, but should it persist when the rest of the system is in place, is there preemptively prepared a PR here than can swap Jest for the ESM-ready [Vitest](https://vitest.dev/):
- #4728

**Comment by HarelM** [comment]
The main problem from my point of view is how to share code between the workers and the main thread.
This is currently relaying on how this is build using commonJs and I don't know how to do it with ESM...
So, if the purpose of ESM is tree shaking, but you get the bundle size duplicated it totally beats the purpose...

**Comment by knd775** [comment]
> So, if the purpose of ESM is tree shaking, but you get the bundle size duplicated it totally beats the purpose...

It increases the package size, but the actual js bundled in an application wouldn't increase with any half-decent bundler

**Comment by jasongitmail** [comment]
@birkskyum Vitest is amazing. It's much faster than Jest and is a good move on its own merit. I switched from Jest to Vitest years ago and never looked back, and it's only going to grow in usage with Evan You now creating [VoidZero](https://voidzero.dev/), which Vitest is part of.

**Comment by birkskyum** [comment]
@jasongitmail , if you find Vitest be _much faster_ already now (before rolldown etc.), could you maybe spare a moment to examine the config of the PR i linked? I'm asking because I haven't managed to figure out why our `test-unit` suite takes for Vitest 2 = 12s, and Jest (we compile with tsc, not swc) = 8s, across mac (locally) and linux/win (CI).

**Comment by jasongitmail** [comment]
@birkskyum I'll drop a comment on the PR.

**Comment by Matb85** [comment]
Hello, did this PR https://github.com/maplibre/maplibre-gl-js/pull/6254 basically add Tree Shaking support?

**Comment by HarelM** [comment]
In theory, in practice you probably can't tree shake a lot of the code. Maybe the UI controls...

**Comment by dangkyokhoang** [comment]
Yes, it seems we probably can't tree shake a lot of code with the v6 esm while having to load a separate worker file.

I tried bundling a minimal app in v 6.0.0-11. V6 is around 930KB + 455KB worker, while v5 is around 1054KB (worker is inlined / shares main code). Then do you have any ideas / plans to decrease the v6 build size?

The way main and worker are shipped in v6 package as separate bundles means we can't by any means configure bundlers (Vite for example) to share code between them. Even with having main and worker both import from a shared module, configuring Vite to split the shared code into a separate chunk that both main and worker can import is not straightforward since Vite by default builds worker separately.

**Comment by HarelM** [comment]
I couldn't find a good esm way to do it (previous code was commonjs with s lot of build hacks to share code). In theory, one cloud create smaller packages of maplibre according to only what's needed, but that's a lot of work and no one decided it's important enough to invest resources in it. So I guess all the talk about bundle size is just a talk...

**Comment by dangkyokhoang** [comment]
Here's a minimal example that has main and worker importing a shared chunk. Bundle size down to 500KB main + 19KB worker + 480KB shared.

https://stackblitz.com/edit/vite-jw7ptvdb?file=main.js

I import maplibre-gl/src directly to let Vite build from source and use a plugin to load and emit the worker in main build so that they share same build context (plugin code from https://github.com/vitejs/vite/issues/16719#issuecomment-2308170706) to allow code splitting to work.

How about maplibre exporting main and worker mjs that both import from a shared chunk? That way users can just import from maplibre package instead of src, and configure their bundler as needed. The current dist files are kept for direct browser imports only.

**Comment by HarelM** [comment]
I don't think that's that easy, but if you are familiar with a solution that will work for all the bundlers tests we have as part of this repo feel free to open a PR. I'll be happy to review it.

**Comment by birkskyum** [comment]
I think we can close this tree-shaking ticket, since tree shaking is now supported after we moved to a ESM-only build. Unused code can be shaken out build-time by bundlers such as vite through named imports - the challenge for the maplibre-gl package is that to keep the full public API available at runtime, it's very limited what can be shaken out build-time, so the gains of doing so are small.

To save further on bundle size, approached like custom builds would have to be pursued instead, where a feature (e.g. heatmap layer) is cut out entirely build-time and thus not available runtime.
